README.md
Seismic
=======

Android device shake detection.



Download
--------

Download [the latest .jar][dl] or depend via Maven:
```xml
<dependency>
  <groupId>com.squareup</groupId>
  <artifactId>seismic</artifactId>
  <version>1.0.3</version>
</dependency>
```
or Gradle:
```groovy
implementation 'com.squareup:seismic:1.0.3'
```

Snapshots of the development version are available in [Sonatype's `snapshots` repository][snap].



License
-------

    Copyright 2012 Square, Inc.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



 [dl]: https://search.maven.org/remote_content?g=com.squareup&a=seismic&v=LATEST
 [snap]: https://oss.sonatype.org/content/repositories/snapshots/


CHANGELOG.md
Change Log
==========

Version 1.0.3 *(2021-09-28)*
----------------------------

 * New: `start()` overload which accepts a sensor start delay. Default is fastest.
 * Fix: Clear sample queue when `stop()` is called. This prevents false positives when restarting.


Version 1.0.2 *(2015-08-04)*
----------------------------

 * Added 'setSensitivity' method for configuring shake sensitivity.


Version 1.0.1 *(2014-12-16)*
----------------------------

 * Remove excessive `Math.sqrt` calculation.


Version 1.0.0 *(2012-09-05)*
----------------------------

Initial release.


library/src/test/java/com/squareup/seismic/ShakeDetectorTest.java
package com.squareup.seismic;

import org.junit.Test;

import java.util.List;

import static org.fest.assertions.api.Assertions.assertThat;

public class ShakeDetectorTest {
  @Test public void testInitialShaking() {
    ShakeDetector.SampleQueue q = new ShakeDetector.SampleQueue();
    assertThat(q.isShaking()).isFalse();
  }

  /** Tests LG Ally sample rate. */
  @Test public void testShakingSampleCount3() {
    ShakeDetector.SampleQueue q = new ShakeDetector.SampleQueue();

    // These times approximate the data rate of the slowest device we've
    // found, the LG Ally.
    // on the LG Ally. The queue holds 500000000 ns (0.5ms) of samples or
    // 4 samples, whichever is greater.
    // 500000000
    q.add(1000000000L, false);
    q.add(1300000000L, false);
    q.add(1600000000L, false);
    q.add(1900000000L, false);
    assertContent(q, false, false, false, false);
    assertThat(q.isShaking()).isFalse();

    // The oldest two entries will be removed.
    q.add(2200000000L, true);
    q.add(2500000000L, true);
    assertContent(q, false, false, true, true);
    assertThat(q.isShaking()).isFalse();

    // Another entry should be removed, now 3 out of 4 are true.
    q.add(2800000000L, true);
    assertContent(q, false, true, true, true);
    assertThat(q.isShaking()).isTrue();

    q.add(3100000000L, false);
    assertContent(q, true, true, true, false);
    assertThat(q.isShaking()).isTrue();

    q.add(3400000000L, false);
    assertContent(q, true, true, false, false);
    assertThat(q.isShaking()).isFalse();
  }

  private void assertContent(ShakeDetector.SampleQueue q, boolean... expected) {
    List<ShakeDetector.Sample> samples = q.asList();

    StringBuilder sb = new StringBuilder();
    for (ShakeDetector.Sample s : samples) {
      sb.append(String.format("[%b,%d] ", s.accelerating, s.timestamp));
    }

    assertThat(samples).hasSize(expected.length);
    for (int i = 0; i < expected.length; i++) {
      assertThat(samples.get(i).accelerating).isEqualTo(expected[i]);
    }
  }

  @Test public void testClear() {
    ShakeDetector.SampleQueue q = new ShakeDetector.SampleQueue();
    q.add(1000000000L, true);
    q.add(1200000000L, true);
    q.add(1400000000L, true);
    assertThat(q.isShaking()).isTrue();
    q.clear();
    assertThat(q.isShaking()).isFalse();
  }

  @Test public void testWindowSize() {
    ShakeDetector.SampleQueue q = new ShakeDetector.SampleQueue();
    q.setWindowSize(4000000000L);
    q.add(1000000000L, true);
    q.add(2000000000L, true);
    q.add(3000000000L, true);
    q.add(4000000000L, true);
    q.add(5000000000L, true);
    q.add(6000000000L, true);
    q.add(7000000000L, true);
    assertThat(q.asList()).hasSize(5);
    q.setWindowSize(1000000000L);
    assertThat(q.asList()).hasSize(4); // MIN_QUEUE_SIZE
  }
}


.buildscript/deploy_snapshot.sh
#!/bin/bash
#
# Deploy a jar, source jar, and javadoc jar to Sonatype's snapshot repo.
#
# Adapted from https://coderwall.com/p/9b_lfq and
# http://benlimmer.com/2013/12/26/automatically-publish-javadoc-to-gh-pages-with-travis-ci/

SLUG="square/seismic"
JDK="oraclejdk8"
BRANCH="master"

set -e

if [ "$TRAVIS_REPO_SLUG" != "$SLUG" ]; then
  echo "Skipping snapshot deployment: wrong repository. Expected '$SLUG' but was '$TRAVIS_REPO_SLUG'."
elif [ "$TRAVIS_JDK_VERSION" != "$JDK" ]; then
  echo "Skipping snapshot deployment: wrong JDK. Expected '$JDK' but was '$TRAVIS_JDK_VERSION'."
elif [ "$TRAVIS_PULL_REQUEST" != "false" ]; then
  echo "Skipping snapshot deployment: was pull request."
elif [ "$TRAVIS_BRANCH" != "$BRANCH" ]; then
  echo "Skipping snapshot deployment: wrong branch. Expected '$BRANCH' but was '$TRAVIS_BRANCH'."
else
  echo "Deploying snapshot..."
  mvn clean source:jar javadoc:jar deploy --settings=".buildscript/settings.xml" -Dmaven.test.skip=true
  echo "Snapshot deployed!"
fi


sample/src/main/java/com/squareup/seismic/sample/Demo.java
package com.squareup.seismic.sample;

import android.app.Activity;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.widget.TextView;
import android.widget.Toast;
import com.squareup.seismic.ShakeDetector;

import static android.view.Gravity.CENTER;
import static android.view.ViewGroup.LayoutParams;
import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;

public class Demo extends Activity implements ShakeDetector.Listener {
  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    SensorManager sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);
    ShakeDetector sd = new ShakeDetector(this);
    sd.start(sensorManager);

    TextView tv = new TextView(this);
    tv.setGravity(CENTER);
    tv.setText("Shake me, bro!");
    setContentView(tv, new LayoutParams(MATCH_PARENT, MATCH_PARENT));
  }

  @Override public void hearShake() {
    Toast.makeText(this, "Don't shake me, bro!", Toast.LENGTH_SHORT).show();
  }
}


library/src/main/java/com/squareup/seismic/ShakeDetector.kt
// Copyright 2010 Square, Inc.
package com.squareup.seismic

import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import java.util.ArrayList

/**
 * Detects phone shaking. If more than 75% of the samples taken in the past 0.5s are
 * accelerating, the device is a) shaking, or b) free falling 1.84m (h =
 * 1/2*g*t^2*3/4).
 *
 * @author Bob Lee (bob@squareup.com)
 * @author Eric Burke (eric@squareup.com)
 */
class ShakeDetector(private val listener: Listener) : SensorEventListener {

    /**
     * When the magnitude of total acceleration exceeds this
     * value, the phone is accelerating.
     */
    private var accelerationThreshold = DEFAULT_ACCELERATION_THRESHOLD

    private val queue = SampleQueue()

    private var sensorManager: SensorManager? = null
    private var accelerometer: Sensor? = null

    /** Listens for shakes.  */
    interface Listener {
        /** Called on the main thread when the device is shaken.  */
        fun hearShake()
    }

    /**
     * Starts listening for shakes on devices with appropriate hardware.
     *
     * @return true if the device supports shake detection.
     */
    fun start(sensorManager: SensorManager): Boolean {
        // Already started?
        if (accelerometer != null) {
            return true
        }

        accelerometer = sensorManager.getDefaultSensor(
                Sensor.TYPE_ACCELEROMETER)

        // If this phone has an accelerometer, listen to it.
        if (accelerometer != null) {
            this.sensorManager = sensorManager
            sensorManager.registerListener(this, accelerometer,
                    SensorManager.SENSOR_DELAY_FASTEST)
        }
        return accelerometer != null
    }

    /**
     * Stops listening.  Safe to call when already stopped.  Ignored on devices
     * without appropriate hardware.
     */
    fun stop() {
        if (accelerometer != null) {
            queue.clear()
            sensorManager!!.unregisterListener(this, accelerometer)
            sensorManager = null
            accelerometer = null
        }
    }

    @Override
    fun onSensorChanged(event: SensorEvent) {
        val accelerating = isAccelerating(event)
        val timestamp = event.timestamp
        queue.add(timestamp, accelerating)
        if (queue.isShaking) {
            queue.clear()
            listener.hearShake()
        }
    }

    /** Returns true if the device is currently accelerating.  */
    private fun isAccelerating(event: SensorEvent): Boolean {
        val ax = event.values[0]
        val ay = event.values[1]
        val az = event.values[2]

        // Instead of comparing magnitude to ACCELERATION_THRESHOLD,
        // compare their squares. This is equivalent and doesn't need the
        // actual magnitude, which would be computed using (expensive) Math.sqrt().
        val magnitudeSquared = (ax * ax + ay * ay + az * az).toDouble()
        return magnitudeSquared > accelerationThreshold * accelerationThreshold
    }

    /** Sets the acceleration threshold sensitivity.  */
    fun setSensitivity(accelerationThreshold: Int) {
        this.accelerationThreshold = accelerationThreshold
    }

    /** Queue of samples. Keeps a running average.  */
    internal class SampleQueue {

        private val pool = SamplePool()

        private var oldest: Sample? = null
        private var newest: Sample? = null
        private var sampleCount: Int = 0
        private var acceleratingCount: Int = 0

        /**
         * Returns true if we have enough samples and more than 3/4 of those samples
         * are accelerating.
         */
        val isShaking: Boolean
            get() = (newest != null
                    && oldest != null
                    && newest!!.timestamp - oldest!!.timestamp >= MIN_WINDOW_SIZE
                    && acceleratingCount >= (sampleCount shr 1) + (sampleCount shr 2))

        /**
         * Adds a sample.
         *
         * @param timestamp    in nanoseconds of sample
         * @param accelerating true if > [.accelerationThreshold].
         */
        fun add(timestamp: Long, accelerating: Boolean) {
            // Purge samples that proceed window.
            purge(timestamp - MAX_WINDOW_SIZE)

            // Add the sample to the queue.
            val added = pool.acquire()
            added.timestamp = timestamp
            added.accelerating = accelerating
            added.next = null
            if (newest != null) {
                newest!!.next = added
            }
            newest = added
            if (oldest == null) {
                oldest = added
            }

            // Update running average.
            sampleCount++
            if (accelerating) {
                acceleratingCount++
            }
        }

        /** Removes all samples from this queue.  */
        fun clear() {
            while (oldest != null) {
                val removed = oldest
                oldest = removed!!.next
                pool.release(removed)
            }
            newest = null
            sampleCount = 0
            acceleratingCount = 0
        }

        /** Purges samples with timestamps older than cutoff.  */
        fun purge(cutoff: Long) {
            while (sampleCount >= MIN_QUEUE_SIZE
                    && oldest != null && cutoff - oldest!!.timestamp > 0) {
                // Remove sample.
                val removed = oldest
                if (removed!!.accelerating) {
                    acceleratingCount--
                }
                sampleCount--

                oldest = removed.next
                if (oldest == null) {
                    newest = null
                }
                pool.release(removed)
            }
        }

        /** Copies the samples into a list, with the oldest entry at index 0.  */
        fun asList(): List<Sample> {
            val list = ArrayList<Sample>()
            var s = oldest
            while (s != null) {
                list.add(s)
                s = s.next
            }
            return list
        }

        companion object {

            /** Window size in ns. Used to compute the average.  */
            private val MAX_WINDOW_SIZE: Long = 500000000 // 0.5s
            private val MIN_WINDOW_SIZE = MAX_WINDOW_SIZE shr 1 // 0.25s

            /**
             * Ensure the queue size never falls below this size, even if the device
             * fails to deliver this many events during the time window. The LG Ally
             * is one such device.
             */
            private val MIN_QUEUE_SIZE = 4
        }
    }

    /** An accelerometer sample.  */
    internal class Sample {
        /** Time sample was taken.  */
        var timestamp: Long = 0

        /** If acceleration > [.accelerationThreshold].  */
        var accelerating: Boolean = false

        /** Next sample in the queue or pool.  */
        var next: Sample? = null
    }

    /** Pools samples. Avoids garbage collection.  */
    internal class SamplePool {
        private var head: Sample? = null

        /** Acquires a sample from the pool.  */
        fun acquire(): Sample {
            var acquired = head
            if (acquired == null) {
                acquired = Sample()
            } else {
                // Remove instance from pool.
                head = acquired.next
            }
            return acquired
        }

        /** Returns a sample to the pool.  */
        fun release(sample: Sample) {
            sample.next = head
            head = sample
        }
    }

    @Override
    fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
    }

    companion object {

        val SENSITIVITY_LIGHT = 11
        val SENSITIVITY_MEDIUM = 13
        val SENSITIVITY_HARD = 15

        private val DEFAULT_ACCELERATION_THRESHOLD = SENSITIVITY_MEDIUM
    }
}


library/src/main/java/com/squareup/seismic/ShakeDetector.java
// Copyright 2010 Square, Inc.
package com.squareup.seismic;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import java.util.ArrayList;
import java.util.List;

/**
 * Detects phone shaking. If more than 75% of the samples taken in the past 0.5s are
 * accelerating, the device is a) shaking, or b) free falling 1.84m (h =
 * 1/2*g*t^2*3/4).
 *
 * @author Bob Lee (bob@squareup.com)
 * @author Eric Burke (eric@squareup.com)
 */
public class ShakeDetector implements SensorEventListener {

  public static final int SENSITIVITY_LIGHT = 11;
  public static final int SENSITIVITY_MEDIUM = 13;
  public static final int SENSITIVITY_HARD = 15;

  private static final int DEFAULT_ACCELERATION_THRESHOLD = SENSITIVITY_MEDIUM;

  /**
   * When the magnitude of total acceleration exceeds this
   * value, the phone is accelerating.
   */
  private int accelerationThreshold = DEFAULT_ACCELERATION_THRESHOLD;

  /** Listens for shakes. */
  public interface Listener {
    /** Called on the main thread when the device is shaken. */
    void hearShake();
  }

  private final SampleQueue queue = new SampleQueue();
  private final Listener listener;

  private SensorManager sensorManager;
  private Sensor accelerometer;

  public ShakeDetector(Listener listener) {
    this.listener = listener;
  }

  /**
   * Starts listening for shakes on devices with appropriate hardware.
   *
   * @return true if the device supports shake detection.
   */
  public boolean start(SensorManager sensorManager) {
    return start(sensorManager, SensorManager.SENSOR_DELAY_FASTEST);
  }

  /**
   * Starts listening for shakes on devices with appropriate hardware.
   * Allowing to set the sensor delay, available values are:
   * SENSOR_DELAY_FASTEST, SENSOR_DELAY_GAME, SENSOR_DELAY_UI, SENSOR_DELAY_NORMAL.
   * @see <a
   * href="https://developer.android.com/reference/android/hardware/SensorManager">SensorManager</a>
   *
   * @return true if the device supports shake detection.
   */
  public boolean start(SensorManager sensorManager, int sensorDelay) {
    // Already started?
    if (accelerometer != null) {
      return true;
    }

    accelerometer = sensorManager.getDefaultSensor(
        Sensor.TYPE_ACCELEROMETER);

    // If this phone has an accelerometer, listen to it.
    if (accelerometer != null) {
      this.sensorManager = sensorManager;
      sensorManager.registerListener(this, accelerometer, sensorDelay);
    }
    return accelerometer != null;
  }

  /**
   * Stops listening.  Safe to call when already stopped.  Ignored on devices
   * without appropriate hardware.
   */
  public void stop() {
    if (accelerometer != null) {
      queue.clear();
      sensorManager.unregisterListener(this, accelerometer);
      sensorManager = null;
      accelerometer = null;
    }
  }

  @Override public void onSensorChanged(SensorEvent event) {
    boolean accelerating = isAccelerating(event);
    long timestamp = event.timestamp;
    queue.add(timestamp, accelerating);
    if (queue.isShaking()) {
      queue.clear();
      listener.hearShake();
    }
  }

  /** Returns true if the device is currently accelerating. */
  private boolean isAccelerating(SensorEvent event) {
    float ax = event.values[0];
    float ay = event.values[1];
    float az = event.values[2];

    // Instead of comparing magnitude to ACCELERATION_THRESHOLD,
    // compare their squares. This is equivalent and doesn't need the
    // actual magnitude, which would be computed using (expensive) Math.sqrt().
    final double magnitudeSquared = ax * ax + ay * ay + az * az;
    return magnitudeSquared > accelerationThreshold * accelerationThreshold;
  }

  /** Sets the acceleration threshold sensitivity. */
  public void setSensitivity(int accelerationThreshold) {
    this.accelerationThreshold = accelerationThreshold;
  }

  /** Sets window size. */
  public void setWindowSize(long size) {
    queue.setWindowSize(size);
  }

  /** Queue of samples. Keeps a running average. */
  static class SampleQueue {

    private static final long DEFAULT_MAX_WINDOW_SIZE = 500000000; // 0.5s
    private static final long DEFAULT_MIN_WINDOW_SIZE = DEFAULT_MAX_WINDOW_SIZE >> 1; // 0.25s

    /** Window size in ns. Used to compute the average. */
    private long maxWindowSize = DEFAULT_MAX_WINDOW_SIZE;
    private long minWindowSize = DEFAULT_MIN_WINDOW_SIZE;

    /**
     * Ensure the queue size never falls below this size, even if the device
     * fails to deliver this many events during the time window. The LG Ally
     * is one such device.
     */
    private static final int MIN_QUEUE_SIZE = 4;

    private final SamplePool pool = new SamplePool();

    private Sample oldest;
    private Sample newest;
    private int sampleCount;
    private int acceleratingCount;

    /**
     * Adds a sample.
     *
     * @param timestamp    in nanoseconds of sample
     * @param accelerating true if > {@link #accelerationThreshold}.
     */
    void add(long timestamp, boolean accelerating) {
      // Purge samples that proceed window.
      purge(timestamp - maxWindowSize);

      // Add the sample to the queue.
      Sample added = pool.acquire();
      added.timestamp = timestamp;
      added.accelerating = accelerating;
      added.next = null;
      if (newest != null) {
        newest.next = added;
      }
      newest = added;
      if (oldest == null) {
        oldest = added;
      }

      // Update running average.
      sampleCount++;
      if (accelerating) {
        acceleratingCount++;
      }
    }

    /** Removes all samples from this queue. */
    void clear() {
      while (oldest != null) {
        Sample removed = oldest;
        oldest = removed.next;
        pool.release(removed);
      }
      newest = null;
      sampleCount = 0;
      acceleratingCount = 0;
    }

    /** Purges samples with timestamps older than cutoff. */
    void purge(long cutoff) {
      while (sampleCount >= MIN_QUEUE_SIZE
          && oldest != null && cutoff - oldest.timestamp > 0) {
        // Remove sample.
        Sample removed = oldest;
        if (removed.accelerating) {
          acceleratingCount--;
        }
        sampleCount--;

        oldest = removed.next;
        if (oldest == null) {
          newest = null;
        }
        pool.release(removed);
      }
    }

    /** Changes window size. */
    void setWindowSize(long size) {
      maxWindowSize = size;
      minWindowSize = maxWindowSize >> 1;
      if (sampleCount > MIN_QUEUE_SIZE) {
        purge(newest.timestamp - maxWindowSize);
      }
    }

    /** Copies the samples into a list, with the oldest entry at index 0. */
    List<Sample> asList() {
      List<Sample> list = new ArrayList<Sample>();
      Sample s = oldest;
      while (s != null) {
        list.add(s);
        s = s.next;
      }
      return list;
    }

    /**
     * Returns true if we have enough samples and more than 3/4 of those samples
     * are accelerating.
     */
    boolean isShaking() {
      return newest != null
          && oldest != null
          && newest.timestamp - oldest.timestamp >= minWindowSize
          && acceleratingCount >= (sampleCount >> 1) + (sampleCount >> 2);
    }
  }

  /** An accelerometer sample. */
  static class Sample {
    /** Time sample was taken. */
    long timestamp;

    /** If acceleration > {@link #accelerationThreshold}. */
    boolean accelerating;

    /** Next sample in the queue or pool. */
    Sample next;
  }

  /** Pools samples. Avoids garbage collection. */
  static class SamplePool {
    private Sample head;

    /** Acquires a sample from the pool. */
    Sample acquire() {
      Sample acquired = head;
      if (acquired == null) {
        acquired = new Sample();
      } else {
        // Remove instance from pool.
        head = acquired.next;
      }
      return acquired;
    }

    /** Returns a sample to the pool. */
    void release(Sample sample) {
      sample.next = head;
      head = sample;
    }
  }

  @Override public void onAccuracyChanged(Sensor sensor, int accuracy) {
  }
}
